relay_test.go

```diff
@@ -1014,3 +1014,159 @@ func TestRelayRaceCompletionAndTimeout(t *testing.T) {
+			msg               string
+			arg2Data          string
+			arg2Flush         bool
+			arg2PostFlushData string
+			noArg3            bool
+			wantEndOffset     int
+			wantHasMore       bool
+		}{
+			{
+				msg:           "all within a frame",
+				arg2Data:      arg2Data,
+				wantEndOffset: wantArg2Start + len(arg2Data),
+				wantHasMore:   false,
+			},
+			{
+				msg:           "arg2 flushed",
+				arg2Data:      arg2Data,
+				arg2Flush:     true,
+				wantEndOffset: wantArg2Start + len(arg2Data),
+				wantHasMore:   true,
+			},
+			{
+				msg:               "arg2 flushed called then write again",
+				arg2Data:          arg2Data,
+				arg2Flush:         true,
+				arg2PostFlushData: "more data",
+				wantEndOffset:     wantArg2Start + len(arg2Data),
+				wantHasMore:       true,
+			},
+			{
+				msg:           "no arg2",
+				wantEndOffset: wantArg2Start,
+				wantHasMore:   false,
+			},
+			{
+				msg:           "XL arg2 which is fragmented",
+				arg2Data:      string(make([]byte, MaxFrameSize+100)),
+				wantEndOffset: wantArg2Start + payloadLeft,
+				wantHasMore:   true,
+			},
+			{
+				msg:           "large arg2 with 3 bytes left for arg3",
+				arg2Data:      string(make([]byte, payloadLeft-3)),
+				wantEndOffset: wantArg2Start + payloadLeft - 3,
+				wantHasMore:   false,
+			},
+			{
+				msg:           "large arg2, 2 bytes left",
+				arg2Data:      string(make([]byte, payloadLeft-2)),
+				wantEndOffset: wantArg2Start + payloadLeft - 2,
+				wantHasMore:   true, // no arg3
+			},
+			{
+				msg:           "large arg2, 2 bytes left, no arg3",
+				arg2Data:      string(make([]byte, payloadLeft-2)),
+				wantEndOffset: wantArg2Start + payloadLeft - 2,
+				noArg3:        true,
+				wantHasMore:   true, // no arg3 and still got CALL_REQ_CONTINUE
+			},
+			{
+				msg:           "large arg2, 1 bytes left",
+				arg2Data:      string(make([]byte, payloadLeft-1)),
+				wantEndOffset: wantArg2Start + payloadLeft - 1,
+				wantHasMore:   true, // no arg3
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.msg, func(t *testing.T) {
+				call, err := client.BeginCall(ctx, ts.HostPort(), ts.ServiceName(), testMethod, nil)
+				require.NoError(t, err, "BeginCall failed")
+				writer, err := call.Arg2Writer()
+				require.NoError(t, err)
+				_, err = writer.Write([]byte(tt.arg2Data))
+				require.NoError(t, err)
+				if tt.arg2Flush {
+					writer.Flush()
+					// tries to write after flush
+					if tt.arg2PostFlushData != "" {
+						_, err := writer.Write([]byte(tt.arg2PostFlushData))
+						require.NoError(t, err)
+					}
+				}
+				require.NoError(t, writer.Close())
+
+				arg3DataToWrite := arg3Data
+				if tt.noArg3 {
+					arg3DataToWrite = ""
+				}
+				require.NoError(t, NewArgWriter(call.Arg3Writer()).Write([]byte(arg3DataToWrite)), "arg3 write failed")
+
+				f := <-inspector.received
+				start := f.Arg2StartOffset()
+				end, hasMore := f.Arg2EndOffset()
+				assert.Equal(t, wantArg2Start, start)
```

Comment:
here and above, add context to the assert so if it fails, we have useful messages